Fabric.Terminology.SqlServer/Persistence/SqlClientTermValueSetRepository.UoWOperation.cs

```diff
@@ -0,0 +1,138 @@
+
+            var batchInsertDtos = codesToAdd.Except(dups).Select(code => ConvertToValueSetCodeDto(valueSetGuid, code)).ToList();
+
+            var removeResult = this.PrepareRemoveCodesOperations(currentCodeDtos, codeDeletes);
+
+            var finalCodes = removeResult.CurrentCodeDtos.Union(batchInsertDtos);
+
+            return new AddRemoveCodeOperations
+            {
+                CurrentCodeDtos = currentCodeDtos,
+                NewCodeDtos = batchInsertDtos,
+                Operations = removeResult.Operations.Union(this.BuildCodeCountOperationList(valueSetGuid, finalCodes))
+                    .ToList()
+            };
+        }
+
+        private IReadOnlyCollection<Operation> BuildRemoveCodesOperationList(
+            IEnumerable<ValueSetCodeDto> originalCodeDtos,
+            IEnumerable<ICodeSystemCode> removeCodes)
+        {
+            var removeGuids = removeCodes.Select(ds => ds.CodeGuid).ToList();
+            return originalCodeDtos.Where(code => removeGuids.Contains(code.CodeGUID.GetValueOrDefault()))
+                .Select(dto => new Operation
+                {
+                    Value = dto,
+                    OperationType = OperationType.Delete
+                }).ToList();
+        }
+
+        private IReadOnlyCollection<Operation> BuildCodeCountOperationList(
+            Guid valueSetGuid,
+            IEnumerable<ValueSetCodeDto> allCodeDtos)
+        {
+            var originalCounts = this.uowManager.GetCodeCountDtos(valueSetGuid);
+            var existingCodeSystems = originalCounts.Select(c => c.CodeSystemGUID);
+
+            var allCodes = allCodeDtos as ValueSetCodeDto[] ?? allCodeDtos.ToArray();
+            var allCodeSystems = allCodes.Select(c => c.CodeSystemGuid).Distinct();
+
+            var recounts = allCodeSystems.Select(
+                            codeSystemGuid => new
+                            {
+                                codeSystemGuid,
+                                valueSetDto = allCodes.First(dto => dto.CodeSystemGuid == codeSystemGuid)
+                            })
+                            .Select(o =>
+                            new ValueSetCodeCountDto(o.valueSetDto.ValueSetGUID,
+                                o.valueSetDto.CodeSystemGuid,
+                                o.valueSetDto.CodeSystemNM,
+                                allCodes.Count(c => c.CodeSystemGuid == o.codeSystemGuid)))
+                                .ToList();
+
+            var operations = recounts.Select(nc =>
+                Maybe.From(originalCounts.FirstOrDefault(ec => ec.CodeSystemGUID == nc.CodeSystemGUID))
+                .Select(
+                    dto =>
+                    {
+                        var op = new Operation();
+                        if (dto.CodeSystemPerValueSetNBR != nc.CodeSystemPerValueSetNBR)
+                        {
+                            dto.CodeSystemPerValueSetNBR = nc.CodeSystemPerValueSetNBR;
```

Comment:
Following Command-Query Separation, we really shouldn't have side-effects happening inside a Select statement. Perhaps this is one of those cases where a traditional `foreach` actually makes more sense.